bigframes/core/nodes.py

```diff
@@ -345,6 +348,67 @@ def transform_children(
         return self
 
 
+## Put ordering in here or just add order_by node above?
+@dataclass(frozen=True)
+class ReadTableNode(BigFrameNode):
+    project_id: str = field()
+    dataset_id: str = field()
+    table_id: str = field()
+
+    physical_schema: Tuple[bq.SchemaField, ...] = field()
+    # Subset of physical schema columns, with chosen BQ types
+    columns: schemata.ArraySchema = field()
+
+    table_session: bigframes.session.Session = field()
+    # Should this even be stored here?
+    # Empty tuple if no primary key (primary key can be any set of columns that together form a unique key)
+    # Empty if no known unique key
+    primary_key: Tuple[str, ...] = field()  # subset of schema
+    # indicates a primary key that is exactly offsets 0, 1, 2, ..., N-2, N-1
+    primary_key_sequential: bool = False
+    snapshot_time: typing.Optional[datetime.datetime] = None
+    # Added for backwards compatibility, not validated
+    sql_predicate: typing.Optional[str] = None
+
+    def __post_init__(self):
+        # enforce invariants
+        physical_names = set(map(lambda i: i.name, self.physical_schema))
+        assert len(self.columns.names) > 0
```

Comment:
Why this assertion? It is possible to create a completely empty table in BQ. Why one would want to do so, I'm not certain, but it is possible.

Reply:
Yeah, I guess we should allow empty tables, removed this constraint